spmd/tensor/redistribute.py

```diff
@@ -111,7 +138,10 @@ def backward(ctx, grad_output: "spmd_tensor.DTensor"):  # type: ignore
         previous_device_mesh = ctx.previous_device_mesh
         return (
             redistribute_spmd_tensor(
-                grad_output, previous_device_mesh, previous_placement
+                grad_output,
+                previous_device_mesh,
+                previous_placement,
+                is_backward=True,
```

Comment:
oh i see this is the backward for redistribute itself. 

I have the feeling that if user is calling redistributed manually, then backward makes sense.

But if we're calling redistribute internally as part of the operator dispatch, we should actually implement that _below_ autograd -- we are redistributing because that specific op implementation doesn't support a specific input sharding. It doesn't say anything about the "backward" implementation for that same op -- if the backward op happens no to support that sharding we should be calling redistribute again anyways. Does that make sense?

Reply:
Yep make sense. This flag and the special treatment is purely for a manual call outside of operator dispatch. For operator dispatch itself, since it's already below autograd, so it will never trigger this `is_backward` flag, and in fact the way I call it in the operator_dispatch is to call the `redistribute_spmd_tensor` api directly instead of calling the autograd function version. https://github.com/pytorch/PiPPy/pull/354/files#diff-19bfa7cfb887150c3ec7d76b0dab34dafc0b97aeeebc2a3cd557d79cffc1d975R69

I tried to reuse the same logic here for manual call outside of operator dispatch and operator call inside torch_dispatch, let me think about how to better model this.

Reply:
Ok I think what i'll do is to just make this special treatment exist in the backward call of the `Redistribute` autograd function, so that in operator_dispatch it will never hit this case.

Reply:
Makes sense, thanks!